### PR TITLE
Sidekiq 7: Add a compatibility layer for `redis-client`

### PR DESCRIPTION
**Pull request by casperisfine**
As discussed in https://github.com/mperham/sidekiq/pull/5253

Switching entirely to `redis-client` is deemed risky, so instead we can support both.

All is needed is a small translation layer, and some very minimal adjustments in the callers.

cc @mperham

### CODE REVIEW

File: lib/sidekiq.rb [hunk 1]
```diff
@@ -117,25 +124,33 @@ def self.redis_info
...
   def self.redis=(hash)
     @redis = if hash.is_a?(ConnectionPool)
       hash
     else
-      Sidekiq::RedisConnection.create(hash)
+      connection_adapter.create(hash)
     end
   end
 
+  def self.base_redis_error
```
Comment (mperham) [inline lib/sidekiq.rb:146]: I don't really want to see this low-level Redis stuff mixed into Sidekiq's top-level Module. Can this go into Sidekiq::Redis* somewhere?

**Review by mperham — COMMENTED** [review]
Wow, this is great work. How do the benchmarks look with the compatibility layer in place?

**Comment by mperham** [comment]
I got to thinking... With a compat layer, we could release this as Sidekiq 6.5 and call the new client "beta" functionality that is not enabled by default and then remove support for the old client in 7.0. WDYT?

**Comment by casperisfine** [comment]
> we could release this as Sidekiq 6.5 and call the new client "beta" functionality that is not enabled by default and then remove support for the old client in 7.0.

Fine by me, you don't even have to commit on the 7.0 removal, you can take the decision later based on feedback.

File: lib/sidekiq.rb (hunk 1 above)
Comment (casperisfine) [reply lib/sidekiq.rb]: I replaced it by `connection_adapter::BaseError` for simplicity. 

But if you find this weird I can do a nested module.

**Comment by casperisfine** [comment]
> How do the benchmarks look with the compatibility layer in place?

I modified the script to make it easier to not add latency

`7-0`
```
byroot@whipstaff:sidekiq:(7-0)$ LATENCY=0 bin/sidekiqload 
2022-04-15T16:00:54.225Z pid=97167 tid=20tf INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T16:00:58.832Z pid=97167 tid=20tf ERROR: Created 500000 jobs
2022-04-15T16:00:58.832Z pid=97167 tid=20tf ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T16:01:23.778Z pid=97167 tid=21z7 ERROR: Done, 500000 jobs in 24.945858 sec
2022-04-15T16:01:23.778Z pid=97167 tid=21z7 ERROR: Now here's the latency for three jobs
0.00034499168395996094
0.00027489662170410156
0.00023293495178222656
byroot@whipstaff:sidekiq:(7-0)$ LATENCY=0 bin/sidekiqload 
2022-04-15T16:01:28.669Z pid=97224 tid=20v8 INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T16:01:32.655Z pid=97224 tid=20v8 ERROR: Created 500000 jobs
2022-04-15T16:01:32.655Z pid=97224 tid=20v8 ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T16:01:57.787Z pid=97224 tid=2210 ERROR: Done, 500000 jobs in 25.132406 sec
2022-04-15T16:01:57.787Z pid=97224 tid=2210 ERROR: Now here's the latency for three jobs
0.0009839534759521484
0.0009448528289794922
0.0003299713134765625
```

This branch with `redis-rb`:

```
byroot@whipstaff:sidekiq:(7-0-dual-client-support)$ LATENCY=0 bin/sidekiqload 
2022-04-15T15:59:09.324Z pid=96727 tid=216r INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T15:59:13.548Z pid=96727 tid=216r ERROR: Created 500000 jobs
2022-04-15T15:59:13.548Z pid=96727 tid=216r ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T15:59:38.887Z pid=96727 tid=21nn ERROR: Done, 500000 jobs in 25.339 sec
2022-04-15T15:59:38.887Z pid=96727 tid=21nn ERROR: Now here's the latency for three jobs
0.0006411075592041016
0.0003681182861328125
0.00021696090698242188
byroot@whipstaff:sidekiq:(7-0-dual-client-support)$ LATENCY=0 bin/sidekiqload 
2022-04-15T15:59:41.035Z pid=96781 tid=20yh INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T15:59:45.318Z pid=96781 tid=20yh ERROR: Created 500000 jobs
2022-04-15T15:59:45.318Z pid=96781 tid=20yh ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T16:00:10.252Z pid=96781 tid=2249 ERROR: Done, 500000 jobs in 24.934286 sec
2022-04-15T16:00:10.252Z pid=96781 tid=2249 ERROR: Now here's the latency for three jobs
0.00045418739318847656
0.00028634071350097656
0.0002338886260986328
```

This branch with `redis-client`:

```
byroot@whipstaff:sidekiq:(7-0-dual-client-support)$ LATENCY=0 SIDEKIQ_REDIS_CLIENT=1 bin/sidekiqload 
2022-04-15T15:55:27.388Z pid=96511 tid=21ez INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T15:55:31.478Z pid=96511 tid=21ez ERROR: Created 500000 jobs
2022-04-15T15:55:31.478Z pid=96511 tid=21ez ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T15:55:54.174Z pid=96511 tid=21u3 ERROR: Done, 500000 jobs in 22.696157 sec
2022-04-15T15:55:54.174Z pid=96511 tid=21u3 ERROR: Now here's the latency for three jobs
0.0003719329833984375
0.0003390312194824219
0.0002541542053222656
byroot@whipstaff:sidekiq:(7-0-dual-client-support)$ LATENCY=0 SIDEKIQ_REDIS_CLIENT=1 bin/sidekiqload 
2022-04-15T15:58:22.975Z pid=96608 tid=214k INFO: Booting Sidekiq 7.0.0 with redis options {:db=>13, :port=>6380}
2022-04-15T15:58:27.773Z pid=96608 tid=214k ERROR: Created 500000 jobs
2022-04-15T15:58:27.773Z pid=96608 tid=214k ERROR: Simulating 0ms of latency between Sidekiq and redis
2022-04-15T15:58:50.282Z pid=96608 tid=21jo ERROR: Done, 500000 jobs in 22.508655 sec
2022-04-15T15:58:50.282Z pid=96608 tid=21jo ERROR: Now here's the latency for three jobs
0.0004181861877441406
0.0002238750457763672
0.0002391338348388672
```

**Comment by casperisfine** [comment]
BTW, not sure if you considered it, but I think it would be interesting to convert this benchmark to `benchmark-ips` to have easier to compare results. e.g. something like `report { push_one_job; process_one_job }`.

**Comment by mperham** [comment]
Sidekiq could always use more benchmarks, I have no problem with anyone reworking or adding things. That said, I prefer a large integration benchmark (i.e. bin/sidekiqload) vs many small microbenchmarks as the former gives a better idea of real world performance.

**Comment by mperham** [comment]
I really don't want ANYTHING in the Sidekiq module if possible. Redis stuff should go in its own namespace. That could be `Sidekiq::RedisConnection`, as far as I'm concerned you do not need to worry about backwards compat with that class as it has always been internal, an impl-detail. It's probably not a good idea to create `Sidekiq::Redis` as that might have namespace collisions but `Sidekiq::RedisClient` is also reasonable.

**Comment by mperham** [comment]
I love what you've done with this pull, Jean. Here's what I'm proposing we do:

1. Since you've implemented a compatibility layer, I will retarget this work to Sidekiq 6.5 as **Beta** functionality, to be released ASAP (Railsconf?).
2. Sidekiq 6.x users and customers can opt into redis-client in order to test in their environment.
3. We'll make redis-client the only option in Sidekiq 7.0 once we have a few months of beta testing.

WDYT?

**Comment by mperham** [comment]
BTW we can't make it the default in 6.x as we support Redis 4.0+ which doesn't have RESP3.

Sidekiq 7 will be Redis 6.2+. Redis 6.2 has been out for over a year now and it is the default version today for both Elasticache and Heroku. Redis 6.2 also deprecates a bunch of commands so it'll be useful for migrating to Redis 7+.

**Comment by casperisfine** [comment]
Sounds good. I’m away for an event this week but I’ll prepare the PRs early next week.

**Pull request closed without merging**